reconstructSPI/iterative_refinement/expectation_maximization.py

```diff
@@ -0,0 +1,524 @@
+                )
+                bayes_factors_2 = IterativeRefinement.compute_bayesian_weights(
+                    particles_f_2[particle_idx], slices_conv_ctfs_2
+                )
+
+                for one_slice_idx in range(bayes_factors_1.shape[0]):
+                    xyz = xyz_rotated[one_slice_idx]
+                    # if this can be vectorized, can avoid loop
+                    inserted_slice_3d_r, count_3d_r = IterativeRefinement.insert_slice(
+                        particle_f_deconv_1.real, xyz, n_pix
+                    )
+                    inserted_slice_3d_i, count_3d_i = IterativeRefinement.insert_slice(
+                        particle_f_deconv_1.imag, xyz, n_pix
+                    )
+                    map_3d_f_updated_1 += inserted_slice_3d_r + 1j * inserted_slice_3d_i
+                    counts_3d_updated_1 += count_3d_r + count_3d_i
+
+                for one_slice_idx in range(bayes_factors_2.shape[0]):
+                    xyz = xyz_rotated[one_slice_idx]
+                    inserted_slice_3d_r, count_3d_r = IterativeRefinement.insert_slice(
+                        particle_f_deconv_2.real, xyz, n_pix
+                    )
+                    inserted_slice_3d_i, count_3d_i = IterativeRefinement.insert_slice(
+                        particle_f_deconv_2.imag, xyz, n_pix
+                    )
+                    map_3d_f_updated_2 += inserted_slice_3d_r + 1j * inserted_slice_3d_i
+                    counts_3d_updated_2 += count_3d_r + count_3d_i
+
+            # apply noise model
+            # half_map_1, half_map_2 come from doing the above
+            # independently. Filter by noise estimate (e.g. multiply
+            # both half maps by FSC)
+            fsc_1d = IterativeRefinement.compute_fsc(
+                map_3d_f_updated_1, map_3d_f_updated_2
+            )
+
+            fsc_3d = IterativeRefinement.expand_1d_to_3d(fsc_1d)
+
+            # multiplicative filter on maps with fsc
+            # The FSC is 1D, one number per spherical shells
+            # it can be expanded back to a multiplicative filter of
+            # the same shape as the maps
+            map_3d_f_filtered_1 = map_3d_f_updated_1 * fsc_3d
+            map_3d_f_filtered_2 = map_3d_f_updated_2 * fsc_3d
+
+            # update iteration
+            half_map_3d_f_1 = map_3d_f_filtered_1
+            half_map_3d_f_2 = map_3d_f_filtered_2
+
+        # final map
+        fsc_1d = IterativeRefinement.compute_fsc(half_map_3d_f_1, half_map_3d_f_2)
+        fsc_3d = IterativeRefinement.expand_1d_to_3d(fsc_1d)
+        map_3d_f_final = (half_map_3d_f_1 + half_map_3d_f_2 / 2) * fsc_3d
+        map_3d_r_final = IterativeRefinement.ifft_3d(map_3d_f_final)
+        half_map_3d_r_1 = IterativeRefinement.ifft_3d(half_map_3d_f_1)
+        half_map_3d_r_2 = IterativeRefinement.ifft_3d(half_map_3d_f_2)
+
+        return map_3d_r_final, half_map_3d_r_1, half_map_3d_r_2, fsc_1d
```

Comment:
I think as this is actually developed, logging will be super useful. So we can think about what to log going forward. Just a heads up now